## Import các thư viện cần thiết

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

---

## Thu thập dữ liệu

Ta dùng bộ dữ liệu được thu thập sẵn, lấy ở trang archive.ics.uci.edu qua đường link [tại đây](https://archive.ics.uci.edu/ml/datasets/Student+Performance).

Ta được phép sử dụng bộ dữ liệu này cho mục đích học tập.

Trích dẫn:
> P. Cortez and A. Silva. Using Data Mining to Predict Secondary School Student Performance. In A. Brito and J. Teixeira Eds., Proceedings of 5th FUture BUsiness TEChnology Conference (FUBUTEC 2008) pp. 5-12, Porto, Portugal, April, 2008, EUROSIS, ISBN 978-9077381-39-7.

Theo mô tả dữ liệu:
> This data approach student achievement in secondary education of two Portuguese schools. The data attributes include student grades, demographic, social and school related features) and it was collected by using school reports and questionnaires. Two datasets are provided regarding the performance in two distinct subjects: Mathematics (mat) and Portuguese language (por). In [Cortez and Silva, 2008], the two datasets were modeled under binary/five-level classification and regression tasks. Important note: the target attribute G3 has a strong correlation with attributes G2 and G1. This occurs because G3 is the final year grade (issued at the 3rd period), while G1 and G2 correspond to the 1st and 2nd period grades. It is more difficult to predict G3 without G2 and G1, but such prediction is much more useful (see paper source for more details).

Trong bài này, vì môn `Mathematics` và `Portuguese` không liên quan đến nhau nên ta chỉ sử dụng bộ dữ liệu thể hiện `Performance Student` trong môn học `Mathematics`

---

## Khám phá dữ liệu

### Đọc dữ liệu từ file csv vào dataframe

In [2]:
mat_df = pd.read_csv("student-mat.csv", delimiter=';')
mat_df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        6   5   6   6  
1      5        3      3     1     1      3        4   5   5   6  
2      4        3      2     2     3      3       10   7   8  10  
3      3        2      2     1     1      5        2  15  14  15  
4      4        3      2     1     2      5        4   6  10  10  

[5 rows x 33 columns]

### Dữ liệu gồm bao nhiều dòng, bao nhiêu cột?

In [3]:
num_rows = mat_df.shape[0]
num_cols = mat_df.shape[1]
print('Số dòng: ', num_rows)
print('Số cột: ', num_cols)

Số dòng:  395
Số cột:  33


### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

Mỗi dòng tương ứng với thông tin (thông tin cá nhân, điểm) của một học sinh tham gia lớp Toán của hai trường được thu thập dữ liệu. Có vẻ như không xảy ra các dòng có ý nghĩa khác nhau

### Dữ liệu có các dòng bị lặp không?

In [4]:
duplicated_rows = mat_df.duplicated().sum()
if duplicated_rows > 0:
    print("Số dòng lặp: ", duplicated_rows)
else:
    print("Không có dòng lặp")

Không có dòng lặp


Vậy là không có dòng lặp bị lặp trong bộ dữ liệu

### Mỗi cột có ý nghĩa gì?

Dưới đây là phần mô tả dữ liệu trong file `student.txt` về các cột trong file dữ liệu "student-mat.csv"

- **school**: student's school (binary: 'GP' - Gabriel Pereira or 'MS' - Mousinho da Silveira)
- **sex**: student's sex (binary: 'F' - female or 'M' - male)
- **age**: student's age (numeric: from 15 to 22)
- **address**: student's home address type (binary: 'U' - urban or 'R' - rural)
- **famsize**: family size (binary: 'LE3' - less or equal to 3 or 'GT3' - greater than 3)
- **Pstatus**: parent's cohabitation status (binary: 'T' - living together or 'A' - apart)
- **Medu**: mother's education (numeric: 0 - none, 1 - primary education (4th grade), 2 – 5th to 9th grade, 3 – secondary education or 4 – higher education)
- **Fedu**: father's education (numeric: 0 - none, 1 - primary education (4th grade), 2 – 5th to 9th grade, 3 – secondary education or 4 – higher education)
- **Mjob**: mother's job (nominal: 'teacher', 'health' care related, civil 'services' (e.g. administrative or police), 'at_home' or 'other')
- **Fjob**: father's job (nominal: 'teacher', 'health' care related, civil 'services' (e.g. administrative or police), 'at_home' or 'other')
- **reason**: reason to choose this school (nominal: close to 'home', school 'reputation', 'course' preference or 'other')
- **guardian**: student's guardian (nominal: 'mother', 'father' or 'other')
- **traveltime**: home to school travel time (numeric: 1 - <15 min., 2 - 15 to 30 min., 3 - 30 min. to 1 hour, or 4 - >1 hour)
- **studytime**: weekly study time (numeric: 1 - <2 hours, 2 - 2 to 5 hours, 3 - 5 to 10 hours, or 4 - >10 hours)
- **failures**: number of past class failures (numeric: n if 1<=n<3, else 4)
- **schoolsup**: extra educational support (binary: yes or no)
- **famsup**: family educational support (binary: yes or no)
- **paid**: extra paid classes within the course subject (Math or Portuguese) (binary: yes or no)
- **activities**: extra-curricular activities (binary: yes or no)
- **nursery**: attended nursery school (binary: yes or no)
- **higher**: wants to take higher education (binary: yes or no)
- **internet**: Internet access at home (binary: yes or no)
- **romantic**: with a romantic relationship (binary: yes or no)
- **famrel**: quality of family relationships (numeric: from 1 - very bad to 5 - excellent)
- **freetime**: free time after school (numeric: from 1 - very low to 5 - very high)
- **goout**: going out with friends (numeric: from 1 - very low to 5 - very high)
- **Dalc**: workday alcohol consumption (numeric: from 1 - very low to 5 - very high)
- **Walc**: weekend alcohol consumption (numeric: from 1 - very low to 5 - very high)
- **health**: current health status (numeric: from 1 - very bad to 5 - very good)
- **absences**: number of school absences (numeric: from 0 to 93)
- **G1**: first period grade (numeric: from 0 to 20)
- **G2**: second period grade (numeric: from 0 to 20)
- **G3**: final grade (numeric: from 0 to 20, output target)

### Kiểu dữ liệu của mỗi cột

In [5]:
col_dtypes = pd.Series(mat_df.dtypes.array, index = mat_df.columns)
col_dtypes 

school        object
sex           object
age            int64
address       object
famsize       object
Pstatus       object
Medu           int64
Fedu           int64
Mjob          object
Fjob          object
reason        object
guardian      object
traveltime     int64
studytime      int64
failures       int64
schoolsup     object
famsup        object
paid          object
activities    object
nursery       object
higher        object
internet      object
romantic      object
famrel         int64
freetime       int64
goout          int64
Dalc           int64
Walc           int64
health         int64
absences       int64
G1             int64
G2             int64
G3             int64
dtype: object

### Xem xét sự phù hợp của dữ liệu

Theo quan sát trong `col_dtypes` các cột `school`,`sex`,`address`,`famsize`,`Pstatus`,`Mjob`,`Fjob`,`reason`,`guardian`,`schoolsup`,`famsup`,
`paid`,`activities`,`nursery`,`higher`,`internet`,`romantic` có kiểu dữ liệu là object

Theo mô tả từ file `student.txt`, các cột này chứa kiểu dữ liệu string, nên ta chuyển tất cả các cột này sang kiểu String

Thật ra, theo file `student.txt` các cột này chứa các binary hoặc nominal, ta có thể chuyển các dữ liệu về dạng số (0,1,2,...) nếu cần xử lý bằng các thuật toán máy học, để trực quan và dễ quan sát (con người có thể đọc được) thì ta chỉ cần chuyển về dạng string

In [20]:
obj = ["school","sex","address","famsize","Pstatus","Mjob","Fjob","reason","guardian","schoolsup","famsup",
     "paid","activities","nursery","higher","internet","romantic"]
mat_df[obj] = mat_df[obj].astype("string")
mat_df.dtypes

school        string
sex           string
age            int64
address       string
famsize       string
Pstatus       string
Medu           int64
Fedu           int64
Mjob          string
Fjob          string
reason        string
guardian      string
traveltime     int64
studytime      int64
failures       int64
schoolsup     string
famsup        string
paid          string
activities    string
nursery       string
higher        string
internet      string
romantic      string
famrel         int64
freetime       int64
goout          int64
Dalc           int64
Walc           int64
health         int64
absences       int64
G1             int64
G2             int64
G3             int64
dtype: object

### Sự phân bố giá trị của mỗi cột

#### Với mỗi cột có kiểu dữ liệu dạng numeric, các giá trị được phân bố như thế nào?

Theo mô tả trong file `student.txt`,  các cột được xem là numeric là `age`,`Medu`,`Fedu`,`traveltime`,`studytime`,`failures`,`famrel`,`freetime`,`goout`,`Dalc`,`Walc`,`health`,`absences`,`G1`,`G2`,`G3`

In [12]:
numerics = ["age","Medu","Fedu","traveltime","studytime","failures","famrel","freetime",
           "goout","Dalc","Walc","health","absences","G1","G2","G3"]

Ta xem xét giá trị thiếu ở các cột

In [17]:
num_missing_ratio = {}
for numeric in numerics:
    num_missing_ratio[numeric] = round(((np.where(pd.isna(mat_df[numeric])==True)[0].size/num_rows)*100),3)
num_missing_ratio

{'age': 0.0,
 'Medu': 0.0,
 'Fedu': 0.0,
 'traveltime': 0.0,
 'studytime': 0.0,
 'failures': 0.0,
 'famrel': 0.0,
 'freetime': 0.0,
 'goout': 0.0,
 'Dalc': 0.0,
 'Walc': 0.0,
 'health': 0.0,
 'absences': 0.0,
 'G1': 0.0,
 'G2': 0.0,
 'G3': 0.0}

Vậy không có giá trị thiếu ở các cột !

Sử dụng hàm describe để xem sự phân bố của các dữ liệu dạng numeric 

In [9]:
mat_df.describe()

age        Medu        Fedu  traveltime   studytime    failures  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean    16.696203    2.749367    2.521519    1.448101    2.035443    0.334177   
std      1.276043    1.094735    1.088201    0.697505    0.839240    0.743651   
min     15.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
25%     16.000000    2.000000    2.000000    1.000000    1.000000    0.000000   
50%     17.000000    3.000000    2.000000    1.000000    2.000000    0.000000   
75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
max     22.000000    4.000000    4.000000    4.000000    4.000000    3.000000   

           famrel    freetime       goout        Dalc        Walc      health  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean     3.944304    3.235443    3.108861    1.481013    2.291139    3.554430   
std      0.896659    0.998862    1.113278    0.890741    1.287897    1.390303   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      4.000000    3.000000    2.000000    1.000000    1.000000    3.000000   
50%      4.000000    3.000000    3.000000    1.000000    2.000000    4.000000   
75%      5.000000    4.000000    4.000000    2.000000    3.000000    5.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

         absences          G1          G2          G3  
count  395.000000  395.000000  395.000000  395.000000  
mean     5.708861   10.908861   10.713924   10.415190  
std      8.003096    3.319195    3.761505    4.581443  
min      0.000000    3.000000    0.000000    0.000000  
25%      0.000000    8.000000    9.000000    8.000000  
50%      4.000000   11.000000   11.000000   11.000000  
75%      8.000000   13.000000   13.000000   14.000000  
max     75.000000   19.000000   19.000000   20.000000

Đa số các cột numeric được chuẩn hóa về các số dạng các choice (1,2,3,...) theo khoảng đã quy định trước, nên `mean` của các cột đó sẽ chỉ có ý nghĩa ở mức tương đối

Ví dụ:
- **Medu**: mother's education (numeric: 0 - none, 1 - primary education (4th grade), 2 - “ 5th to 9th grade, 3 - “ secondary education or 4 - “ higher education)
- Có: `mean` của **Medu** là 2.749367


#### Với mỗi cột có kiểu dữ liệu dạng categorical, các giá trị được phân bố như thế nào?

`mat_df` có các cột catgorical là `school`,`sex`,`address`,`famsize`,`Pstatus`,`Mjob`,`Fjob`,`reason`,`guardian`,`schoolsup`,`famsup`,
`paid`,`activities`,`nursery`,`higher`,`internet`,`romantic`

In [21]:
categorical = obj

Ta xét tỉ lệ giá trị thiếu của các cột

In [23]:
cate_missing_ratio = {}
for cate in categorical:
    cate_missing_ratio[cate] = round(((np.where(pd.isna(mat_df[cate])==True)[0].size/num_rows)*100),3)
cate_missing_ratio

{'school': 0.0,
 'sex': 0.0,
 'address': 0.0,
 'famsize': 0.0,
 'Pstatus': 0.0,
 'Mjob': 0.0,
 'Fjob': 0.0,
 'reason': 0.0,
 'guardian': 0.0,
 'schoolsup': 0.0,
 'famsup': 0.0,
 'paid': 0.0,
 'activities': 0.0,
 'nursery': 0.0,
 'higher': 0.0,
 'internet': 0.0,
 'romantic': 0.0}

Vậy các cột catgorical cũng không có giá trị thiếu nào

Tiếp theo, với mỗi cột categorical, ta xét và lưu vào `cate_col_info_df`
- Số lượng các giá trị (`num_values`)
- Tỉ lệ % (từ 0 đến 100) của mỗi giá trị được sort theo tỉ lệ % giảm dần (`value_percentages`)

In [66]:
cate_col_info_df = pd.DataFrame(index = ['num_values', 'value_percentages'])
for cate in categorical:
    info = pd.Series([mat_df[cate].nunique(), (mat_df[cate].value_counts(normalize=True)*100).to_dict()], 
                  index = ['num_values', 'value_percentages'])
    cate_col_info_df[cate] = info
cate_col_info_df

school  \
num_values                                                         2   
value_percentages  {'GP': 88.35443037974684, 'MS': 11.64556962025...   

                                                                sex  \
num_values                                                        2   
value_percentages  {'F': 52.65822784810127, 'M': 47.34177215189874}   

                                                             address  \
num_values                                                         2   
value_percentages  {'U': 77.72151898734178, 'R': 22.278481012658226}   

                                                             famsize  \
num_values                                                         2   
value_percentages  {'GT3': 71.13924050632912, 'LE3': 28.860759493...   

                                                             Pstatus  \
num_values                                                         2   
value_percentages  {'T': 89.62025316455696, 'A': 10.379746835443038}   

                                                                Mjob  \
num_values                                                         5   
value_percentages  {'other': 35.69620253164557, 'services': 26.07...   

                                                                Fjob  \
num_values                                                         5   
value_percentages  {'other': 54.936708860759495, 'services': 28.1...   

                                                              reason  \
num_values                                                         4   
value_percentages  {'course': 36.708860759493675, 'home': 27.5949...   

                                                            guardian  \
num_values                                                         3   
value_percentages  {'mother': 69.11392405063292, 'father': 22.784...   

                                                           schoolsup  \
num_values                                                         2   
value_percentages  {'no': 87.0886075949367, 'yes': 12.91139240506...   

                                                              famsup  \
num_values                                                         2   
value_percentages  {'yes': 61.26582278481013, 'no': 38.7341772151...   

                                                                paid  \
num_values                                                         2   
value_percentages  {'no': 54.177215189873415, 'yes': 45.822784810...   

                                                          activities  \
num_values                                                         2   
value_percentages  {'yes': 50.8860759493671, 'no': 49.11392405063...   

                                                             nursery  \
num_values                                                         2   
value_percentages  {'yes': 79.49367088607595, 'no': 20.5063291139...   

                                                              higher  \
num_values                                                         2   
value_percentages  {'yes': 94.9367088607595, 'no': 5.063291139240...   

                                                            internet  \
num_values                                                         2   
value_percentages  {'yes': 83.29113924050633, 'no': 16.7088607594...   

                                                            romantic  
num_values                                                         2  
value_percentages  {'no': 66.58227848101265, 'yes': 33.4177215189...